src/backend/gl/src/command.rs

```diff
@@ -326,44 +327,50 @@ impl RawCommandBuffer {
 
     fn update_blend_targets(&mut self, blend_targets: &Vec<pso::ColorBlendDesc>) {
         let max_blend_slots = blend_targets.len();
+        if max_blend_slots == 0 {
+            return;
+        }
 
-        if max_blend_slots > 0 {
-            match self.cache.blend_targets {
-                Some(ref mut cached) => {
-                    if cached.len() < max_blend_slots {
-                        cached.resize(max_blend_slots, None);
-                    }
-                }
-                None => {
-                    self.cache.blend_targets = Some(vec![None; max_blend_slots]);
-                }
-            };
+        if self.cache.blend_targets.len() < max_blend_slots {
+            self.cache.blend_targets.resize(max_blend_slots, None);
         }
 
-        for (slot, blend_target) in blend_targets.iter().enumerate() {
+        let mut all_targets_same = true;
+        for blend_target in &blend_targets[1..] {
+            if blend_target != &blend_targets[0] {
+                all_targets_same = false;
+                break;
+            }
+        }
+
+        if all_targets_same {
             let mut update_blend = false;
-            if let Some(ref mut cached_targets) = self.cache.blend_targets {
-                if let Some(cached_target) = cached_targets.get(slot) {
-                    match cached_target {
-                        &Some(ref cache) => {
-                            if cache != blend_target {
-                                update_blend = true;
-                            }
-                        }
-                        &None => {
-                            update_blend = true;
-                        }
-                    }
+            for cached_target in &mut self.cache.blend_targets {
+                if cached_target.as_ref() != Some(&blend_targets[0]) {
+                    *cached_target = Some(blend_targets[0]);
+                    update_blend = true;
                 }
+            }
+            if update_blend {
+                self.push_cmd(Command::BindAllBlendSlots(blend_targets[0]));
+            }
+        } else {
+            for (slot, blend_target) in blend_targets.iter().enumerate() {
+                let cached_target = self.cache.blend_targets.get_mut(slot).unwrap();
```

Comment:
nit: let's iterate both zipped, i.e. `for (cached, blend_target) = self.cache.blend_targets.iter_mut().zip(blend_targets) {..}`

Reply:
That would be better, except:
1) you still need the slot later so you also have to enumerate (which is easy enough, but it might be a bit big for a single for statement)
2) there's a NLL trick going on where `self.cache.blend_targets` is borrowed for a sub-range of the inner part of the loop.  If I `iter_mut` over it then I borrow part of `self` across `self.push_cmd`.  Ideally we wouldn't, but it wasn't immediately obvious how to do it.

Reply:
I don't think (2) is the case here. For (1) it's fine to split into multiple lines if it becomes long,i.e.
```rust
for (slot, (cached, blend_target)) in self.cache.blend_targets
  .iter_mut()
  .zip(blend_targets.iter())
  .enumerate()
{
///
}
```

Reply:
2 is right, that line you just typed will have a borrowck error because mutable iterator holds a borrow of `blend_targets` across the body of the for loop, which includes the `push_cmd` method call:

```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
   --> src/backend/gl/src/command.rs:368:21
    |
352 |               for (slot, (cached_target, blend_target)) in self.cache.blend_targets
    |                                                            ------------------------
    |                                                            |
    |  __________________________________________________________first mutable borrow occurs here
    | |
353 | |                 .iter_mut()
354 | |                 .zip(blend_targets.iter())
355 | |                 .enumerate()
    | |____________________________- first borrow later used here
...
368 |                       self.push_cmd(Command::BindBlendSlot(slot as _, *blend_target));
    |                       ^^^^ second mutable borrow occurs here

error: aborting due to previous error
```

But yeah, you're right the long for loop is fine, really, other than the borrowck error.

Reply:
One thing, I have no idea why I wrote:

`let cached_target = self.cache.blend_targets.get_mut(slot).unwrap();`

I think it was an artifact of how it was written previously, but in any case I replaced that with the more obvious version.